src/session-pool.ts

```diff
@@ -532,17 +532,22 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
   }
 
   /**
-   * Borrows session from specific group.
+   * Borrows the first session from specific group. This method may only be called if the inventory
+   * actually contains a session of the desired type.
    *
    * @private
    *
    * @param {string} type The desired session type.
    * @return {Session}
    */
   _borrowFrom(type: types): Session {
-    const session = this._inventory[type][0];
-    this._borrow(session);
-    return session;
+    const session = this._inventory[type].pop();
+    if (session) {
```

Comment:
It seems that `borrowFrom` is only called by `_borrowNextAvailableSession` which in turn is only called `_getSession` in cases ether `this.available` is > 0 or on `'available'` event, otherwise a `new Error(errors.Closed)` is thrown.
If so, this step might not be necessary.

WDYT?

Reply:
Yes, you're right about that, but the compiler is not able to see that. So without this check you will get a compiler error. I've added a comment to indicate this.

Reply:
Yes, the come tricky with the type safety monitoring.
I believe the compiler is complaining due to the `undefined` part of the `pop()` typescript return signature of `T | undefined`.
A `!` should be able to take care of that
```ts
    const session = this._inventory[type].pop();

    this._inventory.borrowed.add(session!);
    return session!;
```

Reply:
Ah, yes, that's a lot better. Done.